File: max_vowels.rb
def max_vowels(s, k)
  left = 0
  right = k - 1
  count = 0
  vowels = ['a', 'e', 'i', 'o', 'u']
    
  (left..right).each do |i|
    if(vowels.member?(s[i]))
      count += 1
    end
  end
  
  max_count = count
    
  while right < s.length - 1
    left += 1
    right += 1
 
    count -= 1 if vowels.include?(s[left - 1])
    count += 1 if vowels.include?(s[right])
  
    max_count = [max_count, count].max
  end
  max_count
end

File: max_vowels.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
const vowels = new Set(["a", "e", "i", "o", "u"]);
const isVowel = (char) => {
  return vowels.has(char);
};

var maxVowels = function (s, k) {
  let s_arr = s.split("");
  let l = 0;
  let r = k - 1;
  let initArr = s_arr.slice(l, r + 1);
  let maxCount = 0;

  initArr.forEach((chr) => {
    if (isVowel(chr)) {
      maxCount++;
    }
  });

  let thisCount = maxCount;

  while (r < s.length - 1) {
    l++;
    r++;
    let leftIsVowel = isVowel(s[l - 1]);
    let rightIsVowel = isVowel(s[r]);

    if ((leftIsVowel && rightIsVowel) || (!leftIsVowel && !rightIsVowel)) {
      thisCount = thisCount;
    } else if (leftIsVowel && !rightIsVowel) {
      thisCount = thisCount - 1;
    } else if (!leftIsVowel && rightIsVowel) {
      thisCount++;
    }

    maxCount = Math.max(thisCount, maxCount);
  }

  return maxCount;
};
